### PR TITLE
Implements automated VSIX and NuGet feeds for easy feature dogfooding

### PR DESCRIPTION
**Pull request by kzu**
Implements #45183

Includes commit from #45159 since otherwise it's impossible to test. Once that's merged, I can rebase and exclude it from this PR.

### CODE REVIEW

**Comment by jmarolf** [comment]
If mads and the platform team can confirm that this works now I would love to take this change. Unfortunately, as @jaredpar says, in the past auto-updating tends to lead to corrupted visual studio installations.

here is my understanding of the problem.
1. "Roslyn" is actually 4 vsixes:
  - `VisualStudioSetup.vsix` includes all of the IDE functionality in the editor
  - `VisualStudioInteractiveComponents.vsix` includes the IDE C# REPL
  - `ExpressionEvaluatorPackage.vsix` includes all of the debugging functionality
  - `CompilerExtension.vsix` which is not installed in VS normally but uses some tricks to allow a different compiler to be used from the IDE
2. All of these vsixes need to be _the exact same version_, if any of these 4 are a different versions VS will not work (Roslyn is depended upon bt almost everything in visual studio at this point, if it cannot load VS will just crash in most cases)
1. The VS updater does not do things atomically, so it is very possible for it to update 1 of the 4 extensions and not the others, leaving you with a non-working VS installation. In reality if any of these 4 extensions is not updated you VS installation will be hosed

**Comment by kzu** [comment]
@jmarolf I guess that second one is no longer a thing since #43208, right?

**Comment by jmarolf** [comment]
@kzu I'll need to investigate. From a VS installation perspective, its not true. But our "Roslyn Insiders" vsix still includes it so I would assume that it needs to be kept the same. Unfortunately the VS installer and the VSIX installer do things differently here.

**Comment by jmarolf** [comment]
ah, yes with this PR: https://github.com/dotnet/roslyn/pull/45159 you can disregard the C# REPL

**Comment by CyrusNajmabadi** [comment]
@jmarolf What should we do with this PR?  Thanks! :)

**Comment by jmarolf** [comment]
@CyrusNajmabadi I've assigned this to myself and I'll pull it down and make a determination on Monday. If the nested VSIX problems are resolved then I have no issue with this going in.

File: src/Deployment/RoslynDeployment.csproj [hunk 1]
```diff
@@ -38,14 +45,34 @@
...
+                      Lines="$(UpdateFeedContent)" 
+                      Overwrite="true"
+                      WriteOnlyWhenDifferent="true" />
+    <ItemGroup>
+      <Content Include="$(IntermediateOutputPath)\UpdateFeed.pkgdef" 
+               Link="UpdateFeed.pkgdef"
+               CopyToOutputDirectory="PreserveNewest" 
+               IncludeInVSIX="true" />
+    </ItemGroup>
+  </Target>
```
Comment (sharwell) [inline src/Deployment/RoslynDeployment.csproj:67]: ❗ This doesn't tie in with the necessary up-to-date checks for MSBuild and the IDE

**Comment by sharwell** [comment]
I would not expect this to be reliable or work across arbitrary versions of the IDE. Every time we update a dependency it will have unknown/undefined ramifications on which installed versions can use the new build. The result would be significant confusion for anyone attempting to use this feed, and eventually a strong recommendation that no one do so.

**Comment by kzu** [comment]
That's really fixable by bumping the min required version in the feed/vsixes. You could even automate that by making the min version the same as the VS version used to build in CI.

**Comment by jmarolf** [comment]
> You could even automate that by making the min version the same as the VS version used to build in CI.

Unfortunately, our build doesn't require or use Visual Studio. We import the VSSDK elements via nuget package. We would need some mapping to determine it based on VS package versions alone.

**Pull request closed without merging**